docs/proposals/202004_embedd_cortex_frontend.md

```diff
@@ -0,0 +1,121 @@
+---
+title: Adding a New Thanos Component that Embeds Cortex Query Frontend
+type: proposal
+menu: proposals
+status: in-review
+owner: bwplotka
+---
+
+### Related Tickets
+
+* Response caching: https://github.com/thanos-io/thanos/issues/1651
+* Moving query frontend to separate repo: https://github.com/cortexproject/cortex/issues/1672
+
+## Summary
+
+This proposal describes addition of a new Thanos command (component) into `cmd/thanos` called `frontend`.
+This component will literally import a certain version of Cortex [frontend package](https://github.com/cortexproject/cortex/tree/4410bed704e7d8f63418b02b328ddb93d99fad0b/pkg/querier/frontend).
+
+We will go through rationales, and potential alternatives.
+
+## Motivation
+
+[Cortex Frontend](https://www.youtube.com/watch?v=eyBbImSDOrI&t=2s) was introduced by Tom in August 2019. It was designed
+to be deployed in front of Prometheus Query API in order to ensure:
+
+* Query split by time.
+* Query step alignment.
+* Query retry logic
+* Query limit logic
+* Query response cache in memory, Memcached or Redis.
+
+Since the nature of Cortex backend is really similar to Thanos, with exactly the same PromQL API, and long term capabilities, the caching
+work done for Cortex fits to Thanos. Given also our good collaboration in the past, it feels natural to reuse Cortex's code.
+We even started discussion to move it to separate repo, but there was no motivation towards this, for a good reason.
+At the end we were advertising to use cortex query frontend on production on top of Thanos and this works considerably well, with some
+problems on edge cases and for downsampled data as mentioned [here]()
+
+However, we realized recently that asking users to install suddenly Cortex component on top of Thanos system is extremely confusing:
+
+* Cortex has totally different way of configuring services. It requires to decide what module you have in single YAML file. Thanos in opposite
+have flags and different subcommand for each component.
+* Cortex has bit different way of configuring memcached, which is inconsistent with what we have in Thanos Stote Gateway.
```

Comment:
Cortex memcached client now supports Thanos service discovery too. However, given we're going to vendor it I would advocate to use Thanos memcached client (we just need to make it implement the `cache.Cache` Cortex interface).

Reply:
:+1:

Reply:
Added to action plan